src/memray/_memray/records.h

```diff
@@ -259,8 +261,12 @@ class FrameCollection
     template<typename T>
     auto getIndex(T&& frame) -> std::pair<frame_id_t, bool>
     {
+        std::shared_lock<std::shared_mutex> r_lock(d_mutex);
         auto it = d_frame_map.find(frame);
         if (it == d_frame_map.end()) {
+            r_lock.unlock();
+            std::unique_lock<std::shared_mutex> w_lock(d_mutex);
+
             frame_id_t frame_id =
                     d_frame_map.emplace(std::forward<T>(frame), d_current_frame_id).first->second;
```

Comment:
I believe there's a bug here, actually. We can't ignore the return from `emplace`, because `emplace` can fail if the element we're trying to insert was already inserted by another thread.

Imagine:
- 2 threads both want to insert the same frame
- both successfully grab the shared lock
- both see that the frame doesn't exist yet
- both drop the shared lock
- one grabs the unique lock
- it calls emplace, successfully inserting the element
- the first one drops the unique lock, the second acquires it
- the second one calls emplace, failing to insert the element
- we don't notice, and treat it as though a new ID was registered.